File: frontend/src/shared/api/samples.ts
import { isAxiosError } from 'axios';
import { axiosInstance } from '../config/axios';
import type { PaginatedResponse } from './research';

export interface SampleProtocol {
  id: number;
  test_protocol_number?: string;
  test_protocol_date?: string;
  is_accredited?: boolean;
  formatted_protocol_number?: string;
}

export interface Sample {
  id: number;
  registration_number: string;
  sample_type?: string | null;
  test_object: string;
  sampling_date?: string;
  receiving_date?: string;
  indicators_count: number;
  branch_id?: number;
  sampling_location_id?: number;
  well?: string;
  mode?: string;
  phone?: string;
  selection_conditions?: Record<string, unknown>;
  added_by?: string;
  laboratory_id: number;
  department_id?: number;
  laboratory_name?: string;
  department_name?: string;
  branch_name?: string;
  sampling_location_name?: string;
  protocols?: SampleProtocol[];
  created_at: string;
  updated_at: string;
  deleted_at?: string;
}

export interface SampleCreate {
  registration_number: string;
  sample_type?: string | null;
  test_object: string;
  sampling_date?: string;
  receiving_date?: string;
  indicators_count: number;
  branch_id?: number;
  sampling_location_id?: number;
  well?: string;
  mode?: string;
  phone?: string;
  selection_conditions?: Record<string, unknown>;
  added_by?: string;
  laboratory_id: number;
  department_id?: number;
}

export interface SampleUpdate {
  registration_number?: string;
  sample_type?: string | null;
  test_object?: string;
  sampling_date?: string;
  receiving_date?: string;
  indicators_count: number;
  branch_id?: number;
  sampling_location_id?: number;
  well?: string;
  mode?: string;
  phone?: string;
  selection_conditions?: Record<string, unknown>;
  laboratory_id?: number;
  department_id?: number;
}

export interface SampleFilters {
  registration_number?: string;
  sample_type?: string | null;
  sample_types?: string[];
  test_object?: string;
  test_objects?: string[];
  sampling_location?: string;
  protocols?: string;
  added_by?: string;
  sampling_date_from?: string;
  sampling_date_to?: string;
  receiving_date_from?: string;
  receiving_date_to?: string;
  created_at_from?: string;
  created_at_to?: string;
  [key: string]: unknown;
}

export interface SelectionConditionsField {
  variable: string;
  unit: string;
}

const parseContentDispositionFilename = (contentDisposition: string): string | null => {
  const utf8NameMatch = contentDisposition.match(/filename\*\s*=\s*UTF-8''([^;]+)/i);
  const basicNameMatch = contentDisposition.match(/filename\s*=\s*"?([^"]+)"?/i);
  const rawName = utf8NameMatch?.[1] ?? basicNameMatch?.[1];
  if (!rawName) {
    return null;
  }
  try {
    return decodeURIComponent(rawName);
  } catch {
    return rawName;
  }
};

const buildSampleListParams = (
  page: number | undefined,
  pageSize: number | undefined,
  filters: SampleFilters | undefined,
  sorting: { sort_by?: string; sort_order?: 'asc' | 'desc' } | undefined,
  laboratoryId: number | undefined,
  departmentId: number | undefined
): Record<string, unknown> => {
  const params: Record<string, unknown> = {};

  if (page !== undefined) {
    params.page = page;
  }

  if (pageSize !== undefined) {
    params.page_size = pageSize;
  }

  if (laboratoryId) {
    params.laboratory_id = laboratoryId;
  }

  if (departmentId) {
    params.department_id = departmentId;
  }

  if (filters?.registration_number) {
    params.search = filters.registration_number;
  }

  if (filters?.sample_types && filters.sample_types.length > 0) {
    params.sample_types = filters.sample_types;
  } else if (filters?.sample_type) {
    params.sample_type = filters.sample_type;
  }

  if (filters?.test_objects && filters.test_objects.length > 0) {
    params.test_objects = filters.test_objects;
  } else if (filters?.test_object) {
    params.test_object = filters.test_object;
  }

  if (filters?.sampling_location) {
    params.search_sampling_location = filters.sampling_location;
  }

  if (filters?.protocols) {
    params.search_protocols = filters.protocols;
  }

  if (filters?.added_by) {
    params.search_added_by = filters.added_by;
  }

  if (filters?.sampling_date_from) {
    params.sampling_date_from = filters.sampling_date_from;
  }
  if (filters?.sampling_date_to) {
    params.sampling_date_to = filters.sampling_date_to;
  }
  if (filters?.receiving_date_from) {
    params.receiving_date_from = filters.receiving_date_from;
  }
  if (filters?.receiving_date_to) {
    params.receiving_date_to = filters.receiving_date_to;
  }
  if (filters?.created_at_from) {
    params.created_at_from = filters.created_at_from;
  }
  if (filters?.created_at_to) {
    params.created_at_to = filters.created_at_to;
  }

  if (sorting?.sort_by) {
    params.sort_by = sorting.sort_by;
    params.sort_order = sorting.sort_order || 'desc';
  }

  return params;
};

export const samplesApi = {
  getSampleTypes: async (): Promise<string[]> => {
    const response = await axiosInstance.get<string[]>('/api/sample-types/');
    return response.data;
  },

  getTestObjects: async (laboratoryId?: number, departmentId?: number): Promise<string[]> => {
    const params: Record<string, number> = {};

    if (laboratoryId) {
      params.laboratory_id = laboratoryId;
    }

    if (departmentId) {
      params.department_id = departmentId;
    }

    const response = await axiosInstance.get<string[]>('/api/test-objects/names/', {
      params,
    });
    return response.data;
  },

  getSamples: async (
    page?: number,
    pageSize?: number,
    filters?: SampleFilters,
    sorting?: { sort_by?: string; sort_order?: 'asc' | 'desc' },
    laboratoryId?: number,
    departmentId?: number
  ): Promise<PaginatedResponse<Sample>> => {
    const params = buildSampleListParams(
      page,
      pageSize,
      filters,
      sorting,
      laboratoryId,
      departmentId
    );

    const response = await axiosInstance.get<PaginatedResponse<Sample>>('/api/samples/', {
      params,
    });
    return response.data;
  },

  exportSamples: async (
    filters?: SampleFilters,
    sorting?: { sort_by?: string; sort_order?: 'asc' | 'desc' },
    laboratoryId?: number,
    departmentId?: number
  ): Promise<{ blob: Blob; filename: string; total: number | null }> => {
    const params = buildSampleListParams(
      undefined,
      undefined,
      filters,
      sorting,
      laboratoryId,
      departmentId
    );

    try {
      const response = await axiosInstance.get<Blob>('/api/samples/export/', {
        params,
        responseType: 'blob',
      });

      let filename = 'Поступления_проб.xlsx';
      const contentDisposition =
        response.headers['content-disposition'] || response.headers['Content-Disposition'] || '';
      const parsedFilename = parseContentDispositionFilename(String(contentDisposition));
      if (parsedFilename) {
        filename = parsedFilename;
      }

      const totalHeader = response.headers['x-export-total'] || response.headers['X-Export-Total'];
      const total =
        totalHeader !== undefined && totalHeader !== null && totalHeader !== ''
          ? parseInt(String(totalHeader), 10)
          : null;

      return { blob: response.data, filename, total };
    } catch (error) {
      if (
        isAxiosError(error) &&
        error.response?.status === 400 &&
        error.response.data instanceof Blob
      ) {
        const text = await error.response.data.text();
        let detail = 'Нет данных для экспорта по выбранным фильтрам';
        try {
          const body = JSON.parse(text) as { detail?: string };
          if (body.detail) {
            detail = body.detail;
          }
        } catch {
          // ответ не JSON — оставляем сообщение по умолчанию
        }
        throw new Error(detail);
      }
      throw error;
    }
  },

  getSample: async (id: number): Promise<Sample> => {
    const response = await axiosInstance.get<Sample>(`/api/samples/${id}/`);
    return response.data;
  },

  createSample: async (data: SampleCreate): Promise<Sample> => {
    const response = await axiosInstance.post<Sample>('/api/samples/', data);
    return response.data;
  },

  updateSample: async (id: number, data: SampleUpdate): Promise<Sample> => {
    const response = await axiosInstance.patch<Sample>(`/api/samples/${id}/`, data);
    return response.data;
  },

  deleteSample: async (id: number): Promise<void> => {
    await axiosInstance.delete(`/api/samples/${id}/`);
  },

  getSelectionConditionsFields: async (
    laboratoryId: number,
    departmentId?: number
  ): Promise<SelectionConditionsField[]> => {
    const params: Record<string, unknown> = {
      laboratory_id: laboratoryId,
    };

    if (departmentId) {
      params.department_id = departmentId;
    }

    const response = await axiosInstance.get<
      PaginatedResponse<{ conditions: SelectionConditionsField[] }>
    >('/api/selection-conditions/', { params });

    if (response.data.items.length > 0) {
      return response.data.items[0].conditions;
    }

    return [];
  },

  getRegistrationNumbers: async (
    laboratoryId?: number,
    departmentId?: number,
    methodId?: number,
    search?: string
  ): Promise<{ samples: Sample[] }> => {
    const params: Record<string, unknown> = {};

    if (laboratoryId) {
      params.laboratory_id = laboratoryId;
    }

    if (departmentId) {
      params.department_id = departmentId;
    }

    if (methodId) {
      params.method_id = methodId;
    }

    if (search) {
      params.search = search;
    }

    const response = await axiosInstance.get<{ samples: Sample[] }>(
      '/api/get-registration-numbers/',
      { params }
    );
    return response.data;
  },
};

File: backend/api/samples.py
from __future__ import annotations
from typing import get_args
from fastapi import APIRouter, Depends, Query
from core.deps import DbSession, SampleListFilters, UserPermissions
from core.exceptions import ValidationError
from core.responses import build_attachment_response
from schemas.pagination import PaginatedResponse
from schemas.sample import (
    SAMPLE_TYPE_CHOICES,
    SampleCreate,
    SampleResponse,
    SampleUpdate,
)
from services.access_control import enforce_nav_access, enforce_samples_mutation
from services.sample import (
    build_sample_response,
    build_samples_list_response,
    create_sample,
    delete_sample,
    get_registration_number_samples,
    get_sample_response_data,
    get_samples,
    require_sample_by_id,
    update_sample,
)
from services.samples_excel import build_samples_export_excel

router = APIRouter()


@router.get(
    "/sample-types/",
    response_model=list[str],
    summary="Получение списка типов проб",
    description="Возвращает список доступных типов проб.",
    responses={200: {"description": "Список типов проб успешно получен"}},
)
# @IsAuthenticated
async def get_sample_types(effective: UserPermissions):
    """Возвращает список доступных типов проб."""
    return list(get_args(SAMPLE_TYPE_CHOICES))


@router.get(
    "/samples/",
    response_model=PaginatedResponse[SampleResponse],
    summary="Получение списка проб",
    description=(
        "Возвращает список проб с пагинацией или без. "
        "Если page и page_size не указаны, возвращает все записи. "
        "Поддерживает фильтрацию по лабораториям и подразделениям, поиск и сортировку."
    ),
    responses={200: {"description": "Список проб успешно получен"}},
)
# @IsAuthenticated
async def list_samples(
    db: DbSession,
    effective: UserPermissions,
    filters: SampleListFilters = Depends(),
):
    """Возвращает список проб с пагинацией или без."""
    enforce_nav_access(effective, "samples", filters.laboratory_id, filters.department_id)
    samples, total, total_pages = await get_samples(
        db,
        laboratory_id=filters.laboratory_id,
        department_id=filters.department_id,
        page=filters.page,
        page_size=filters.page_size,
        search=filters.search,
        search_sampling_location=filters.search_sampling_location,
        search_protocols=filters.search_protocols,
        search_added_by=filters.search_added_by,
        sample_type=filters.sample_type,
        sample_types=filters.sample_types,
        test_object=filters.test_object,
        test_objects=filters.test_objects,
        sort_by=filters.sort_by,
        sort_order=filters.sort_order,
        sampling_date_from=filters.sampling_date_from,
        sampling_date_to=filters.sampling_date_to,
        receiving_date_from=filters.receiving_date_from,
        receiving_date_to=filters.receiving_date_to,
        created_at_from=filters.created_at_from,
        created_at_to=filters.created_at_to,
    )

    items = await build_samples_list_response(db, samples)

    return PaginatedResponse(
        items=items,
        total=total,
        page=filters.page if filters.page is not None else 1,
        page_size=filters.page_size if filters.page_size is not None else total,
        total_pages=total_pages,
    )


@router.get(
    "/samples/export/",
    summary="Формирование файла экспорта таблицы поступления проб",
    description=(
        "Формирует xlsx-файл со всеми пробами по текущим фильтрам и сортировке, включая расчеты и методы исследований."
    ),
    responses={
        200: {"description": "Файл Excel успешно сформирован"},
        400: {"description": "Нет данных для экспорта"},
    },
)
# @IsAuthenticated
async def export_samples(
    db: DbSession,
    effective: UserPermissions,
    filters: SampleListFilters = Depends(),
):
    """Формирует xlsx-файл таблицы поступления проб по текущим фильтрам и сортировке."""
    enforce_nav_access(effective, "samples", filters.laboratory_id, filters.department_id)
    excel_bytes, total = await build_samples_export_excel(
        db,
        laboratory_id=filters.laboratory_id,
        department_id=filters.department_id,
        search=filters.search,
        search_sampling_location=filters.search_sampling_location,
        search_protocols=filters.search_protocols,
        search_added_by=filters.search_added_by,
        sample_type=filters.sample_type,
        sample_types=filters.sample_types,
        test_object=filters.test_object,
        test_objects=filters.test_objects,
        sort_by=filters.sort_by,
        sort_order=filters.sort_order,
        sampling_date_from=filters.sampling_date_from,
        sampling_date_to=filters.sampling_date_to,
        receiving_date_from=filters.receiving_date_from,
        receiving_date_to=filters.receiving_date_to,
        created_at_from=filters.created_at_from,
        created_at_to=filters.created_at_to,
    )

    if total == 0 or not excel_bytes:
        raise ValidationError("Нет данных для экспорта по выбранным фильтрам")

    filename = "Поступления_проб.xlsx"
    return build_attachment_response(
        excel_bytes,
        filename,
        "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        extra_headers={"X-Export-Total": str(total)},
    )


@router.post(
    "/samples/",
    response_model=SampleResponse,
    status_code=201,
    summary="Добавление новой пробы",
    description="Добавляет новую пробу на основе переданных данных.",
    responses={
        201: {"description": "Проба успешно добавлена"},
        400: {"description": "Некорректные данные для добавления пробы"},
    },
)
# @IsAuthenticated
async def create_sample_endpoint(
    sample_data: SampleCreate,
    db: DbSession,
    effective: UserPermissions,
):
    """Добавляет новую пробу на основе переданных данных."""
    enforce_nav_access(effective, "samples", sample_data.laboratory_id, sample_data.department_id)
    sample = await create_sample(db, sample_data)
    return await get_sample_response_data(db, sample.id)


@router.get(
    "/samples/{sample_id}/",
    response_model=SampleResponse,
    summary="Получение пробы по ID",
    description="Возвращает информацию о пробе по ее идентификатору.",
    responses={
        200: {"description": "Проба успешно получена"},
        404: {"description": "Проба не найдена"},
    },
)
# @IsAuthenticated
async def get_sample(
    sample_id: int,
    db: DbSession,
    effective: UserPermissions,
):
    """Возвращает информацию о пробе по ее идентификатору."""
    sample = await require_sample_by_id(db, sample_id)
    enforce_nav_access(effective, "samples", sample.laboratory_id, sample.department_id)
    return await get_sample_response_data(db, sample_id)


@router.patch(
    "/samples/{sample_id}/",
    response_model=SampleResponse,
    summary="Обновление пробы",
    description="Обновляет существующую пробу.",
    responses={
        200: {"description": "Проба успешно обновлена"},
        404: {"description": "Проба не найдена"},
    },
)
# @IsAuthenticated
async def update_sample_endpoint(
    sample_id: int,
    sample_data: SampleUpdate,
    db: DbSession,
    effective: UserPermissions,
):
    """Обновляет существующую пробу."""
    sample = await require_sample_by_id(db, sample_id)
    enforce_samples_mutation(effective, "update", sample.laboratory_id, sample.department_id)
    await update_sample(db, sample_id, sample_data)
    return await get_sample_response_data(db, sample_id)


@router.delete(
    "/samples/{sample_id}/",
    status_code=204,
    summary="Удаление пробы",
    description="Выполняет мягкое удаление пробы.",
    responses={
        204: {"description": "Проба успешно удалена"},
        404: {"description": "Проба не найдена"},
    },
)
# @IsAuthenticated
async def delete_sample_endpoint(
    sample_id: int,
    db: DbSession,
    effective: UserPermissions,
):
    """Выполняет мягкое удаление пробы."""
    sample = await require_sample_by_id(db, sample_id)
    enforce_samples_mutation(effective, "delete", sample.laboratory_id, sample.department_id)
    await delete_sample(db, sample_id)


@router.get(
    "/get-registration-numbers/",
    response_model=dict,
    summary="Поиск регистрационных номеров проб по методу исследования",
    description=(
        "Возвращает список проб с расчетами по указанному методу исследования, "
        "которые соответствуют поисковому запросу. Используется для автодополнения "
        "при вводе регистрационного номера пробы."
    ),
    responses={200: {"description": "Список проб успешно получен"}},
)
# @IsAuthenticated
async def get_registration_numbers(
    db: DbSession,
    effective: UserPermissions,
    laboratory_id: int | None = Query(None),
    department_id: int | None = Query(None),
    method_id: int | None = Query(None),
    search: str | None = Query(None),
):
    """Возвращает список проб с расчетами по указанному методу исследования."""
    enforce_nav_access(effective, "samples", laboratory_id, department_id)
    if not method_id:
        return {"samples": []}

    samples = await get_registration_number_samples(db, method_id, laboratory_id, department_id, search)

    return {"samples": [build_sample_response(sample) for sample in samples]}
